### ISSUE TITLE
statement 'cout<<endl<<"App Running"<<endl;' causes wdt & cpu-reset in endless loop, with several different backtraces. Also happens with standard esp-idf lcd/lvgl example (IDFGH-7728)

### ISSUE THREAD

**Issue by moefear85** [espressif/esp-idf#9269]
## Environment

- Module or chip used:  esp32-d0wdq6 revision 1
- IDF version (run ``git describe --tags`` to find it): v5.0-c2ccc38
- Build System:   idf.py
- Operating System:    |Linu
- Using an IDE?: vscode
- Power Supply:        external 3.3V

## Problem Description

When using cout as shown in the code below, the app crashes in an endless loop without resetting, and displays the errors shown in the logs endlessly, every couple of seconds. There are two different variants of backtrace errors that get produced every time I run the code. No error occurs if I replace cout with a corresponding printf statement.

### Code to reproduce this issue

```cpp
#include <iostream>
#include "esp_event.h"

using std::endl;
using std::cout;

extern "C" void app_main(void)
{
    esp_event_loop_create_default();

    cout<<endl<<"App Running"<<endl;
}

```
## Debug Logs

```
W (171) boot.esp32: PRO CPU has been reset by WDT.
W (172) boot.esp32: WDT reset info: PRO CPU PC=0x4008870a
0x4008870a: _xt_context_save at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/freertos/FreeRTOS-Kernel/portable/xtensa/xtensa_context.S:157

W (172) boot.esp32: WDT reset info: APP CPU PC=0x400882f2
0x400882f2: panic_handler at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/esp_system/port/panic_handler.c:139 (discriminator 1)

W (171) boot.esp32: PRO CPU has been reset by WDT.
W (172) boot.esp32: WDT reset info: PRO CPU PC=0x400886f8
0x400886f8: _xt_context_save at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/freertos/FreeRTOS-Kernel/portable/xtensa/xtensa_context.S:138

W (172) boot.esp32: WDT reset info: APP CPU PC=0x400882f2
0x400882f2: panic_handler at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/esp_system/port/panic_handler.c:139 (discriminator 1)
```
as well as:

```
--- idf_monitor on /dev/ttyUSB1 2073600 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---

W (171) boot.esp32: PRO CPU has been reset by WDT.
W (172) boot.esp32: WDT reset info: PRO CPU PC=0x4008870a
0x4008870a: spi_flash_protected_read_mmu_entry at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/spi_flash/flash_mmap.c:309

W (172) boot.esp32: WDT reset info: APP CPU PC=0x400882f2
0x400882f2: spi_flash_disable_interrupts_caches_and_other_cpu at /home/moefear/source/esp-idf-v5.0-c2ccc38/components/spi_flash/cache_utils.c:150 (discriminator 1)
```

## Other items if possible
[debug.log](https://github.com/espressif/esp-idf/files/9029452/debug.log)
[debug2.log](https://github.com/espressif/esp-idf/files/9029453/debug2.log)
[sdkconfig.txt](https://github.com/espressif/esp-idf/files/9029455/sdkconfig.txt)

**Comment by moefear85** [comment]
I just tested the example in:

> $IDF_PATH/examples/peripherals/lcd/i80_controller/

Irrespective of the example configuration, irrespective whether PSRAM is attached or not (tried wroom-32 and wrover-32 modules), irrespective if an LCD is actually attached or I'm using a bare module, irrespective which chip I use, I also get the same errors (see log) endlessly.

I just re-cloned to make sure there is nothing wrong with my esp-idf copy. The issue persists. I'm on c2ccc38 without ANY modifications.

[lvgl.log](https://github.com/espressif/esp-idf/files/9032807/lvgl.log)

**Comment by 0xjakob** [comment]
@moefear85 I'm not able to reproduce this. I tried on a WROOM module and the output there looks normal:
```
...
I (0) cpu_start: Starting scheduler on APP CPU.

App Running
```

What is weird is that the registers aren't printed normally on your backtrace, even though it's enabled in the sdkconfig file. Maybe you can try to enable GDB-stub and then print the backtrace, if that works for you.

What you can also try is re-installing the toolchain by deleting the respective toolchain from the `.espressif` folder. The download will be quite large, do this only if you have good internet.

We furthermore noticed that using `std::ostream` will create a lot of code. Hence, we try to avoid it. This advice will probably go into a future C++ programming guide for IDF, too. Hence, if you can, we'd recommend to switch to `printf()` anyway.

**Comment by moefear85** [comment]
@0xjakob I'll try a full fresh install then first

**Comment by 0xjakob** [comment]
@moefear85 Sorry, we already reproduced the issue (when using your sdkconfig). We'll take a look.

**Comment by moefear85** [comment]
ah ok thx... i just finished getting everything ready, was going test. i'll check back later then

**Comment by 0xjakob** [comment]
@moefear85 Have you checked the UART and SPI settings? They are heavily modified (frequency is 80 MHz, mode is QIO), I'm not sure if this works e.g. on a normal development board. What configuration are you using? (module, flash, UART)?

When I don't use your configuration (i.g. removing the sdkconfig, then building again), then everything works as expected.

**Comment by moefear85** [comment]
@0xjakob 

I always use 80 MHz flash & QIO for all projects, i've never run into any issues that would be resolved by lowering them. Moreover, the lcd example uses a vanilla sdkconfig yet the problem still occurs.
What I don't remember at the moment, is whether the problem is specific to esp-idf v5, but i'll check and respond.

either way i'll test both projects using standard spi/flash setings & respond. i'll also check which modules i'm using.

But if you assume freq/qio is the problem, then you or espressif or anyone else with known high quality modules/flash should be able to get them working at those settings, hence confirm easily whether it is a freq/quality issue, no?

**Comment by 0xjakob** [comment]
> But if you assume freq/qio is the problem, then you or espressif or anyone else with known high quality modules/flash should be able to get them working at those settings, hence confirm easily whether it is a freq/quality issue, no?

Yes, hence me asking about the configuration (I mean hardware configuration here). Because when I tested on an ESP32 WROVER module and vanilla config, then everything worked fine, so I assumed it's the configuration that has some issues.

**Comment by moefear85** [comment]
@0xjakob 

OK I found the trigger of the problem, but not the source nor the solution. So, to systematically reproduce the problem:

Start with any fresh empty project. Insert the code into main.c as mentioned at the top (the event related statements aren't necessary/relevant) after renaming it to main.cpp. Then in sdkconfig, simply set the default log verbosity to "warning".

and believe it or not that's it. that, in combination with the cout statement, is triggering the problem. (it happens every few seconds, like 5 to 10 seconds, so don't assume it's working if it doesn't immediately crash after booting).

**Comment by 0xjakob** [comment]
@moefear85 Thanks for the heads-up! I could reproduce the issue now (wouldn't have thought simply increasing the warning level would lead to this). I'll take a deeper look.

I also talked to our chip support. You're right, 80MHz and QIO *should* be supported on all our official modules. The reason it's not the default is that it's the safest option for pairing a custom flash chip with an ESP chip.

**Comment by moefear85** [comment]
no prob, glad to contribute what i can

**Comment by SoucheSouche** [comment]
Hi @moefear85,
After a little bit of digging, it seems that the problem is triggered by the call to
```
std::cout << std::endl;
```
For instance, just adding a character like 
```
std::cout << ' ' << std::endl;
```
seems to make the problem disappear.
More precisely, the implementation of std::endl is done as follows
```
flush(os.put(os.widen('\n')
```
And it seems that the call to cout.put() before a call to the operator << is causing the core to hang. After a couple of second the task watchdog is triggered and the processor is restarted.

I will investigate more deeply the effect of this function in the IDF environment and will let you know later on if I managed to isolate the reason why the core hangs.

**Comment by SoucheSouche** [comment]
Hi @moefear85,

After quite some debugging we managed to isolate the reason why printing a single character when the log level is set to warning causes the core to hang and the board to restart. It turns out that the function path when calling a `printf` or `cout <<` with only one character was changed when our toolchain was updated (several months ago). The new function path end up calling functions that cause a stack corruption leading to an interrupt watchdog reset.

A structure in the new toolchain was changed from 32bits to 64bits but as some functions placed in ROM are using this structure expecting it to be 32bits hence the stack corruption.

The problem affects all boards and the solution reside in preventing all calls to those ROM functions relying on this modified structure.

I will keep you posted on the progress of the fix but it seems that it might take some time to resolved it.

In the meantime I would suggest (if not already done) replacing:
```
cout<<endl<<"App Running"<<endl;
```
with:
```
cout<<"App Running"<<endl;
```
as a temporary fix in your project.
